File: uikit/src/main/java/com/sendbird/uikit/widgets/ImageWaffleView.java
package com.sendbird.uikit.widgets;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Shader;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.appcompat.widget.AppCompatImageView;

import com.sendbird.uikit.R;
import com.sendbird.uikit.SendBirdUIKit;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class ImageWaffleView extends ViewGroup {
    protected final int DIVIDER_WIDTH = 1;
    private final int ROUNDING_RADIUS = 1;
    private static final int ROUND_BORDER = 1;

    private final Paint roundingPaint = new Paint();
    private final Paint borderPaint = new Paint();
    private final RectF canvasBounds = new RectF();
    private final Canvas tempCanvas = new Canvas();
    private Bitmap tempBitmap;

    private static class KillerWaffleChildImageView extends AppCompatImageView {

        final ImageWaffleView imageWaffleView;

        KillerWaffleChildImageView(ImageWaffleView imageWaffleView) {
            super(imageWaffleView.getContext());

            this.imageWaffleView = imageWaffleView;

            setScaleType(ImageView.ScaleType.CENTER_CROP);
        }

        @Override
        public void requestLayout() {
            // suppress layout for performance
            super.requestLayout();
            imageWaffleView.forceLayout();
        }

        @Override
        public void invalidate() {
            super.invalidate();
            imageWaffleView.invalidate();
        }
    }

    public ImageWaffleView(Context context) {
        this(context, null);
    }

    public ImageWaffleView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ImageWaffleView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        roundingPaint.setAntiAlias(true);
        roundingPaint.setFilterBitmap(true);
        borderPaint.setAntiAlias(true);
    }

    protected ImageView prepareSingleImageView() {
        return prepareImageViews(1).get(0);
    }

    protected List<ImageView> prepareImageViews(int length) {
        if (length > 4 || length < 0) {
            throw new IllegalArgumentException("Invalid length : " + length);
        }

        Queue<ImageView> prevImageViews = new LinkedList<>();
        if (getChildCount() == length) {
            for (int i = 0; i < getChildCount(); i++) {
                prevImageViews.add((ImageView) getChildAt(i));
            }
        } else {
            removeAllViews();
        }

        List<ImageView> prepared = new ArrayList<>(length);

        for (int i = 0; i < length; i++) {
            prepared.add(pollOrNewImageView(prevImageViews));
        }

        ImageView toBeRemoved;
        while ((toBeRemoved = prevImageViews.poll()) != null) {
            removeView(toBeRemoved);
        }

        return Collections.unmodifiableList(prepared);
    }

    private ImageView pollOrNewImageView(Queue<ImageView> prevImageViews) {
        ImageView polled = prevImageViews.poll();
        if (polled != null) {
            return polled;
        }

        ImageView imageView = new KillerWaffleChildImageView(this);
        addView(imageView);
        return imageView;
    }

    @Override
    protected void dispatchDraw(Canvas canvas) {
        int width = getWidth();
        int height = getHeight();

        if (isInEditMode()) {
            Paint debugPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
            debugPaint.setColor(getResources().getColor(R.color.background_400));
            canvas.drawRoundRect(new RectF(0, 0, width, height), ROUNDING_RADIUS, ROUNDING_RADIUS, debugPaint);
            return;
        }

        if (width == 0 || height == 0) {
            super.dispatchDraw(canvas);
            return;
        }

        if (tempBitmap == null || tempBitmap.isRecycled() || (tempBitmap.getWidth() != width || tempBitmap.getHeight() != height)) {
            if (tempBitmap != null) {
                tempBitmap.recycle();
            }

            try {
                tempBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
            } catch (OutOfMemoryError e) {
                tempBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565);
            }
        } else {
            tempBitmap.eraseColor(Color.TRANSPARENT);
        }
        tempCanvas.setBitmap(tempBitmap);

        //drawBackground(tempCanvas); // draw below children
        super.dispatchDraw(tempCanvas); // draw children
        //drawForeground(tempCanvas); // draw above children
        //drawGlass(tempCanvas); // draw top-most layer

        roundingPaint.setShader(new BitmapShader(tempBitmap, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP));

        int paddingLeft = getPaddingLeft();
        int paddingRight = getPaddingRight();
        int paddingTop = getPaddingTop();
        int paddingBottom = getPaddingBottom();
        canvasBounds.set(0, 0, width, height);
        borderPaint.setColor(getResources().getColor(SendBirdUIKit.isDarkMode() ? R.color.ondark_04 : R.color.onlight_04));

        canvas.drawRoundRect(canvasBounds, (float)(width / 2), (float)(height / 2), borderPaint);
        canvasBounds.set(ROUND_BORDER + paddingLeft, ROUND_BORDER + paddingTop, width - ROUND_BORDER - paddingRight, height - ROUND_BORDER - paddingBottom);
        canvas.drawRoundRect(canvasBounds, (float)(width / 2), (float)(height / 2), roundingPaint);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int specWidth = MeasureSpec.getSize(widthMeasureSpec);
        int specHeight = MeasureSpec.getSize(heightMeasureSpec);

        int halfWidth = (specWidth - DIVIDER_WIDTH + ROUND_BORDER) / 2;
        int halfHeight = (specHeight - DIVIDER_WIDTH + ROUND_BORDER) / 2;

        switch (getChildCount()) {
            case 1:
                measureInGrid(getChildAt(0), specWidth, specHeight);
                break;
            case 2:
                measureInGrid(getChildAt(0), halfWidth, specHeight);
                measureInGrid(getChildAt(1), halfWidth, specHeight);
                break;
            case 3:
                measureInGrid(getChildAt(0), specWidth, halfHeight);
                measureInGrid(getChildAt(1), halfWidth, halfHeight);
                measureInGrid(getChildAt(2), halfWidth, halfHeight);
                break;
            case 4:
                measureInGrid(getChildAt(0), halfWidth, halfHeight);
                measureInGrid(getChildAt(1), halfWidth, halfHeight);
                measureInGrid(getChildAt(2), halfWidth, halfHeight);
                measureInGrid(getChildAt(3), halfWidth, halfHeight);
                break;
        }

        setMeasuredDimension(specWidth, specHeight);
    }

    private void measureInGrid(View view, int width, int height) {
        view.measure(MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY), MeasureSpec.makeMeasureSpec(height, MeasureSpec.EXACTLY));
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        int halfWidth = (getMeasuredWidth() - DIVIDER_WIDTH - ROUND_BORDER) / 2;
        int halfHeight = (getMeasuredHeight() - DIVIDER_WIDTH - ROUND_BORDER) / 2;

        switch (getChildCount()) {
            case 1:
                getChildAt(0).layout(0, 0, getMeasuredWidth(), getMeasuredHeight());
                break;
            case 2:
                getChildAt(0).layout(0, 0, halfWidth, getMeasuredHeight());
                getChildAt(1).layout(getMeasuredWidth() - halfWidth, 0, getMeasuredWidth(), getMeasuredHeight());
                break;
            case 3:
                getChildAt(0).layout(0, 0, getMeasuredWidth(), halfHeight);
                getChildAt(1).layout(0, getMeasuredHeight() - halfHeight, halfWidth, getMeasuredHeight());
                getChildAt(2).layout(getMeasuredWidth() - halfWidth, getMeasuredHeight() - halfHeight, getMeasuredWidth(), getMeasuredHeight());
                break;
            case 4:
                getChildAt(0).layout(0, 0, halfWidth, halfHeight);
                getChildAt(1).layout(getMeasuredWidth() - halfWidth, 0, getMeasuredWidth(), halfHeight);
                getChildAt(2).layout(0, getMeasuredHeight() - halfHeight, halfWidth, getMeasuredHeight());
                getChildAt(3).layout(getMeasuredWidth() - halfWidth, getMeasuredHeight() - halfHeight, getMeasuredWidth(), getMeasuredHeight());
                break;
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();

        if (tempBitmap != null) {
            tempBitmap.recycle();
        }
    }
}

File: uikit/src/main/java/com/sendbird/uikit/internal/ui/widgets/ImageWaffleView.kt
package com.sendbird.uikit.internal.ui.widgets

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapShader
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Shader
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.appcompat.widget.AppCompatImageView
import androidx.core.content.ContextCompat
import com.sendbird.uikit.R
import com.sendbird.uikit.SendbirdUIKit
import java.util.Collections
import java.util.LinkedList
import java.util.Queue

internal open class ImageWaffleView @JvmOverloads internal constructor(
    context: Context,
    attrs: AttributeSet? = null,
) : ViewGroup(context, attrs) {

    companion object {
        private const val ROUND_BORDER = 1
        private const val DIVIDER_WIDTH = 1
    }

    private val roundingPaint = Paint()
    private val borderPaint = Paint()
    private val canvasBounds = RectF()
    private val tempCanvas = Canvas()
    private var tempBitmap: Bitmap? = null

    private class KillerWaffleChildImageView constructor(val imageWaffleView: ImageWaffleView) :
        AppCompatImageView(imageWaffleView.context) {
        override fun requestLayout() {
            // suppress layout for performance
            super.requestLayout()
            imageWaffleView.forceLayout()
        }

        override fun invalidate() {
            super.invalidate()
            imageWaffleView.invalidate()
        }

        init {
            scaleType = ScaleType.CENTER_CROP
        }
    }

    init {
        roundingPaint.isAntiAlias = true
        roundingPaint.isFilterBitmap = true
        borderPaint.isAntiAlias = true
    }

    open fun prepareSingleImageView(): ImageView {
        return prepareImageViews(1)[0]
    }

    open fun prepareImageViews(length: Int): List<ImageView> {
        require(!(length > 4 || length < 0)) { "Invalid length : $length" }
        val prevImageViews: Queue<ImageView> = LinkedList()
        if (childCount == length) {
            for (i in 0 until childCount) {
                prevImageViews.add(getChildAt(i) as ImageView)
            }
        } else {
            removeAllViews()
        }
        val prepared: MutableList<ImageView> = ArrayList(length)
        for (i in 0 until length) {
            prepared.add(pollOrNewImageView(prevImageViews))
        }
        var toBeRemoved: ImageView?
        while (prevImageViews.poll().also { toBeRemoved = it } != null) {
            removeView(toBeRemoved)
        }
        return Collections.unmodifiableList(prepared)
    }

    private fun pollOrNewImageView(prevImageViews: Queue<ImageView>): ImageView {
        val polled = prevImageViews.poll()
        if (polled != null) {
            return polled
        }
        val imageView: ImageView = KillerWaffleChildImageView(this)
        addView(imageView)
        return imageView
    }

    override fun dispatchDraw(canvas: Canvas) {
        val width = width
        val height = height
        if (isInEditMode) {
            val debugPaint = Paint(Paint.ANTI_ALIAS_FLAG)
            debugPaint.color = ContextCompat.getColor(context, R.color.background_400)
            val ROUNDING_RADIUS = 1
            canvas.drawRoundRect(
                RectF(0F, 0F, width.toFloat(), height.toFloat()),
                ROUNDING_RADIUS.toFloat(),
                ROUNDING_RADIUS.toFloat(),
                debugPaint
            )
            return
        }
        if (width == 0 || height == 0) {
            super.dispatchDraw(canvas)
            return
        }

        tempBitmap?.let {
            if (it.isRecycled || it.width != width || it.height != height) {
                it.recycle()
                tempBitmap = try {
                    Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
                } catch (e: OutOfMemoryError) {
                    Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565)
                }
            } else {
                it.eraseColor(Color.TRANSPARENT)
            }
        } ?: run {
            tempBitmap = try {
                Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
            } catch (e: OutOfMemoryError) {
                Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565)
            }
        }

        tempCanvas.setBitmap(tempBitmap)

        // draw below children
        // drawBackground(tempCanvas);

        // draw children
        super.dispatchDraw(tempCanvas)

        // draw above children
        // drawForeground(tempCanvas);

        // draw top-most layer
        // drawGlass(tempCanvas);
        tempBitmap?.let {
            roundingPaint.shader =
                BitmapShader(it, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP)
        }

        val paddingStart = paddingStart
        val paddingEnd = paddingEnd
        val paddingTop = paddingTop
        val paddingBottom = paddingBottom
        canvasBounds.set(0f, 0f, width.toFloat(), height.toFloat())
        borderPaint.color =
            ContextCompat.getColor(context, if (SendbirdUIKit.isDarkMode()) R.color.ondark_text_disabled else R.color.onlight_text_disabled)
        canvas.drawRoundRect(
            canvasBounds,
            (width / 2).toFloat(),
            (height / 2).toFloat(),
            borderPaint
        )
        canvasBounds.set(
            ROUND_BORDER + paddingStart.toFloat(),
            ROUND_BORDER + paddingTop.toFloat(),
            (width - ROUND_BORDER - paddingEnd).toFloat(),
            (height - ROUND_BORDER - paddingBottom).toFloat()
        )
        canvas.drawRoundRect(
            canvasBounds,
            (width / 2).toFloat(),
            (height / 2).toFloat(),
            roundingPaint
        )
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val specWidth = MeasureSpec.getSize(widthMeasureSpec)
        val specHeight = MeasureSpec.getSize(heightMeasureSpec)
        val halfWidth = (specWidth - DIVIDER_WIDTH + ROUND_BORDER) / 2
        val halfHeight = (specHeight - DIVIDER_WIDTH + ROUND_BORDER) / 2
        when (childCount) {
            1 -> measureInGrid(getChildAt(0), specWidth, specHeight)
            2 -> {
                measureInGrid(getChildAt(0), halfWidth, specHeight)
                measureInGrid(getChildAt(1), halfWidth, specHeight)
            }
            3 -> {
                measureInGrid(getChildAt(0), specWidth, halfHeight)
                measureInGrid(getChildAt(1), halfWidth, halfHeight)
                measureInGrid(getChildAt(2), halfWidth, halfHeight)
            }
            4 -> {
                measureInGrid(getChildAt(0), halfWidth, halfHeight)
                measureInGrid(getChildAt(1), halfWidth, halfHeight)
                measureInGrid(getChildAt(2), halfWidth, halfHeight)
                measureInGrid(getChildAt(3), halfWidth, halfHeight)
            }
        }
        setMeasuredDimension(specWidth, specHeight)
    }

    private fun measureInGrid(view: View, width: Int, height: Int) {
        view.measure(
            MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY),
            MeasureSpec.makeMeasureSpec(height, MeasureSpec.EXACTLY)
        )
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        val halfWidth = (measuredWidth - DIVIDER_WIDTH - ROUND_BORDER) / 2
        val halfHeight = (measuredHeight - DIVIDER_WIDTH - ROUND_BORDER) / 2
        when (childCount) {
            1 -> getChildAt(0).layout(0, 0, measuredWidth, measuredHeight)
            2 -> {
                getChildAt(0).layout(0, 0, halfWidth, measuredHeight)
                getChildAt(1).layout(measuredWidth - halfWidth, 0, measuredWidth, measuredHeight)
            }
            3 -> {
                getChildAt(0).layout(0, 0, measuredWidth, halfHeight)
                getChildAt(1).layout(0, measuredHeight - halfHeight, halfWidth, measuredHeight)
                getChildAt(2).layout(
                    measuredWidth - halfWidth, measuredHeight - halfHeight,
                    measuredWidth, measuredHeight
                )
            }
            4 -> {
                getChildAt(0).layout(0, 0, halfWidth, halfHeight)
                getChildAt(1).layout(measuredWidth - halfWidth, 0, measuredWidth, halfHeight)
                getChildAt(2).layout(0, measuredHeight - halfHeight, halfWidth, measuredHeight)
                getChildAt(3).layout(
                    measuredWidth - halfWidth, measuredHeight - halfHeight,
                    measuredWidth, measuredHeight
                )
            }
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        tempBitmap?.recycle()
    }
}
